server/svix-server/src/queue/redis.rs

```diff
@@ -36,89 +48,222 @@ use super::{
+                    e.kind(),
+                    e.detail(),
+                    e.code(),
+                    e.category(),
+                    e
+                )
+            };
+        }
+    }
+
+    // Redis durationns are given in integer numbers of milliseconds, so the pending_duration (the
+    // time in which a task is allowed to be processing before being restarted) must be converted to
+    // one.
+    let pending_duration: i64 = pending_duration
+        .as_millis()
+        .try_into()
+        .expect("Pending duration out of bounds");
+
     let worker_pool = pool.clone();
+
+    // This is the background thread that monitors the DELAYED queue for messages ready to be
+    // inserted into the MAIN queue and that monitors the pending tasks of the MAIN queue for
+    // messages that must be reinserted.
     tokio::spawn(async move {
-        // FIXME: enforce we only have one such worker (locking)
+        // FIXME: enforce we only have one such worker via locking
         let batch_size: isize = 50;
-        let task_validity_duration = chrono::Duration::from_std(TASK_VALIDITY_DURATION).unwrap();
+        let pool = worker_pool;
 
+        // Before entering the loop, migrate v1 queues to v2 and v2 queues to v3.
         {
-            let pool = worker_pool.clone();
             let mut pool = pool.get().await.unwrap();
 
             // drain legacy queues:
-            migrate_legacy_queues(&mut pool).await;
+            migrate_v1_to_v2_queues(&mut pool).await;
+            migrate_v2_to_v3_queues(&mut pool).await;
         }
 
         loop {
-            let mut pool = worker_pool.get().await.unwrap();
+            let mut pool = pool.get().await.unwrap();
 
+            // First look for delayed keys whose time is up and add them to the main qunue
             let timestamp = Utc::now().timestamp();
             let keys: Vec<String> = pool
-                .zrangebyscore_limit(DELAYED, 0isize, timestamp, 0isize, batch_size)
+                .zrangebyscore_limit(delayed_queue_name, 0isize, timestamp, 0isize, batch_size)
                 .await
                 .unwrap();
             if !keys.is_empty() {
                 // FIXME: needs to be a transaction
-                let keys: Vec<(String, String)> =
-                    pool.zpopmin(DELAYED, keys.len() as isize).await.unwrap();
-                let keys: Vec<String> = keys
-                    .into_iter()
-                    .map(|x| x.0)
-                    .map(|x| regenerate_key(&x))
+                let keys: Vec<(String, String)> = pool
+                    .zpopmin(delayed_queue_name, keys.len() as isize)
+                    .await
+                    .unwrap();
+                let tasks: Vec<&str> = keys
+                    .iter()
+                    // All information is stored in the key in which the ID and JSON formated task
+                    // are separated by a `|`. So, take the key, then take the part after the `|`
```

Comment:
Isn't the ID external to the value, and the value is just there on its own as the data of key `data`?

Reply:
The delayed queue is untouched as of now, so they're exactly the same

Reply:
Sweet.